### DISCUSSION TITLE
db session middleware with custom  exception middleware raises AnyIO EndOfSteam when client closes early

### DISCUSSION THREAD

**Discussion by s-quinn** [fastapi/fastapi#8582] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import time
from typing import Callable, Awaitable

import uvicorn
from fastapi import FastAPI
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker, declarative_base
from starlette.requests import Request
from starlette.responses import Response



SQLALCHEMY_DATABASE_URL = "postgresql://user:password@postgresserver/db"
engine = create_engine(
    SQLALCHEMY_DATABASE_URL
)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
Base = declarative_base(bind=engine)

app = FastAPI()

class ExceptionHandlerMiddleware:
    async def __call__(
        self,
        request: Request,
        call_next: Callable[[Request], Awaitable[Response]],
    ) -> Response:
        try:
            return await call_next(request)
        except Exception as exc:
            print(exc)


app.middleware('http')(ExceptionHandlerMiddleware())


@app.middleware("http")
async def db_session_middleware(request: Request, call_next):
    response = Response("Internal server error", status_code=500)
    try:
        request.state.db = SessionLocal()
        response = await call_next(request)
    finally:
        request.state.db.close()
    return response


@app.get('/')
def test():
    time.sleep(2)
    return {'message': 'hello'}


if __name__ == '__main__':
    uvicorn.run(
        'example:app',
        reload=True,
    )
```

### Description

We are using the db session middleware exactly as described [here](https://fastapi.tiangolo.com/tutorial/sql-databases/?h=sql#create-a-middleware) in combination with a custom http exception middleware. When a client is disconnecting from us early we get spammed with a bunch of `No response returned` errors from starlette. 

How to reproduce: 

Open browser and call the endpoint `/` 
Close the window before the request completes. 
You will see that AnyIO raises a `EndOfStream` error which ends up being propogated as a `Runtime Error` with the message as `No response returned`

### Operating System

Linux, macOS

### FastAPI Version

0.73.0

### Python Version

Python 3.9.6

### Additional Context

SQLAlchemy==1.4.31

**Comment by teamhide** [comment]
+1

**Comment by wwarne** [comment]
+1 too.

**Comment by guya-pecan** [comment]
+1 as well

**Comment by leydenberg** [comment]
+1 :(

**Comment by robotadam** [comment, marked as answer]
I did some tests on this after Starlette merged https://github.com/encode/starlette/pull/1715, which contains further fixes for this. It resolved the issues in my fastapi app with early disconnection, so I'm feeling positive that the pending update will be the solution for this bug. https://github.com/tiangolo/fastapi/pull/5471
